hydra-node/test/Hydra/NodeSpec.hs

```diff
@@ -0,0 +1,327 @@
+module Hydra.NodeSpec where
+
+import Cardano.Crypto.DSIGN (SigDSIGN (SigMockDSIGN))
+import Control.Exception (ErrorCall)
+import Data.Sequence (Seq ((:|>)))
+import qualified Data.Sequence as Seq
+import Hydra.Chain (Chain (Chain), OnChainTx (..))
+import Hydra.ClientInput (ClientInput (..))
+import Hydra.HeadLogic (
+  Environment (Environment),
+  Event (..),
+  HeadState (..),
+  LogicError (InvalidSnapshot),
+  SnapshotStrategy (..),
+ )
+import Hydra.Ledger (Party, Signed (UnsafeSigned), SigningKey, Tx, deriveParty)
+import Hydra.Ledger.Simple (SimpleTx (..), simpleLedger, utxoRef, utxoRefs)
+import Hydra.Network (Network (..))
+import Hydra.Network.Message (Message (..))
+import qualified Hydra.Network.Message as Msg
+import Hydra.Node (HydraNode (..), createEventQueue, createHydraHead, processNextEvent)
+import Hydra.Prelude
+import Test.Hspec (Selector, Spec, describe)
+import Test.Hspec.QuickCheck (prop)
+import Test.QuickCheck (Property)
+import Test.QuickCheck.Monadic (assert, monadicIO, run)
+
+spec :: Spec
+spec =
+  describe "Hydra Node" $
+    prop
+      "throws an error given next snapshot is acked before current snapshot finishes"
+      prop_unexpectedSnapshotDoesNotThrowError
+
+prop_unexpectedSnapshotDoesNotThrowError :: Events -> Property
+prop_unexpectedSnapshotDoesNotThrowError (Events evs) = monadicIO $ do
+  result <- run $ do
+    node <- createHydraNode 20 [30, 10] SnapshotAfterEachTx
+    feedEvents node (prefix <> toList evs)
+
+  assert $ isNotErrorCall result
+
+isNotErrorCall :: Maybe (FeedError SimpleTx) -> Bool
+isNotErrorCall Nothing = True
+isNotErrorCall (Just (FeedError [])) = False
+isNotErrorCall (Just (FeedError _)) = True
+
+newtype Events = Events (Seq (Event SimpleTx))
+  deriving (Eq, Show)
+
+instance Arbitrary Events where
+  arbitrary = pure (Events $ Seq.fromList reducedEvents)
+
+  shrink = \case
```

Comment:
@KtorZ I wrote this basic shrinker for the events stream, did you have something more involved in mind? I really want now to have that state-machine based testing in place so that we can generate events and feed that to quickcheck to cover the corner cases of our protocol. While you are away enjoying packing stuff and moving furniture, I want to work on a QSM based test for this whole logic.